lib/oauth2-client.js

```diff
@@ -0,0 +1,185 @@
+ *
+ * @param {object} options - Options to use.
+ * @param {string} options.client_id - The ID of the client.
+ * @param {string} options.client_secret - The client secret.
+ * @param {string} options.token_endpoint - The URL to call.
+ * @param {string} options.grant_type - The grant type.
+ * @param {string} options.scope - The scope.
+ * @param {object} options.httpsAgent - A NodeJS HTTPS Agent (`https.Agent`).
+ * @param {number} options.maxRetries - The maximum number of times to retry
+ *  the request.
+ *
+ * @returns {string} The accessToken.
+ */
+async function getNewAccessToken({
+  client_id, client_secret, token_endpoint, grant_type, scope, httpsAgent,
+  maxRetries = 3
+}) {
+  const body = new URLSearchParams();
+  body.set('client_id', client_id);
+  body.set('client_secret', client_secret);
+  body.set('grant_type', grant_type);
+  body.set('scope', scope);
+
+  let response;
+  const token = {client_id, client_secret};
+
+  while(maxRetries >= 0) {
+    response = await httpCall({url: token_endpoint, body, httpsAgent});
+    if(response && response.access_token) {
+      token.access_token = response.access_token;
+      return token;
+    }
+    --maxRetries;
+  }
+  throw new BedrockError(
+    `Service Unavilable`, 'InvalidStateError', {
+      httpStatusCode: 503,
+      public: true
+    });
+}
+
+async function httpCall({url, body, httpsAgent}) {
+  let response;
+  try {
+    ({data: response} = await httpClient.post(url, {
+      body,
+      agent: httpsAgent
+    }));
+  } catch(e) {
+    logger.error('Error getting accessToken.', {error: e});
+  }
+  if(response && response.access_token) {
+    return response;
+  }
+  return false;
+}
+
+/**
+ * Inserts a token into the database.
+ *
+ * @param {object} options - Options to use.
+ * @param {object} options.token - The token record, including client id and
+ * accessToken.
+ *
+ * @returns {string} The token record.
+ */
+export async function insertToken({token}) {
+  const now = Date.now();
+  const meta = {created: now, updated: now};
+  // expires after 2 weeks
+  const ttl = 2 * 7 * 24 * 60 * 60 * 1000;
+  token.expires = new Date(now + ttl);
+  let record = {
+    meta,
+    token
+  };
+  const collection = database.collections['access-tokens'];
+  try {
+    const result = await collection.insertOne(record, database.writeOptions);
+    record = result.ops[0];
+  } catch(e) {
+    if(!database.isDuplicateError(e)) {
+      throw e;
+    }
+  }
+  return record;
+}
+
+/**
+ * Queries the database for an existing accessToken for the client id.
+ *
+ * @param {object} options - Options to use.
+ * @param {string} options.client_id - The ID of the client.
+ *
+ * @returns {string} The accessToken if there is one, null if not.
+ */
+export async function checkDatabase({client_id}) {
+  const query = {
+    'token.client_id': client_id
+  };
+  let token;
+  const result = await database.collections['access-tokens'].findOne(query);
```

Comment:
The naming convention we use for database records is `record`. If you make a mongodb call and it returns something other than the record itself, then `result` is a good name. But `findOne` actually just returns the `record` directly, I believe -- with no wrapper that has other result information.

```suggestion
  const record = await database.collections['access-tokens'].findOne(query);
```